Mathlib/Tactic/Linter/Header.lean

```diff
@@ -0,0 +1,281 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang, Damiano Testa
+-/
+import Lean.Elab.Command
+
+/-!
+#  The "header" linter
+
+The "header" style linter checks that a file starts with
+```
+/-
+Copyright ...
+Apache ...
+Authors ...
+-/
+
+import*
+module doc-string*
+rest
+```
+It emits a warning if
+* the copyright statement is malformed;
+* `Mathlib.Tactic` is imported;
+* any import in `Lake` is present;
+* the first non-`import` command is not a module doc-string.
+
+The linter allows `import`-only files and does not require a copyright statement in `Mathlib.Init`.
+
+The strategy used by the linter is as follows.
+The linter computes the end position of the first module doc-string of the file,
+resorting to the end of the file, if there is no module doc-string.
+Next, the linter tries to parse the file up to the position determined above.
+
+If the parsing is successful, the linter checks the resulting `Syntax` and behaves accordingly.
+
+If the parsing is not successful, this already means that there is some "problematic" command
+after the imports.
+In particular, there is a command that is not a doc-module string immediately following the last.
+import: the file should be flagged by the linter.
+Hence, the linter then falls back to parsing the header of the file adding a spurious `section`
+after.
+This makes it possible for the linter to check all the header of the file, emit warnings that
+could arise from this part and also flag that the file should contain a doc-module string after
+the `import` statements.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+`firstNonImport? stx` assumes that the input `Syntax` is of kind `Lean.Parser.Module.module`.
+It returns
+* `none`, if `stx` consists only of `import` statements,
+* the first non-`import` command in `stx`, otherwise.
+
+The intended use-case is to use the output of `testParseModule` as the input of
+`firstNonImport?`.
+-/
+def firstNonImport? : Syntax → Option Syntax
+  | .node _ ``Lean.Parser.Module.module #[_header, .node _ `null args] => args[0]?
+  | _=> some .missing  -- this is unreachable, if the input comes from `testParseModule`
+
+/-- `getImportIds s` takes as input `s : Syntax`.
```

Comment:
Maybe I could, but that would not give me the `Syntax` (in particular, the position information) of the imports, right?  I think that I use that to log warnings when necessary.

Reply:
Fair point. Another good argument could be to avoid importing `ImportGraph` in `Mathlib.Init`.
What do you think; would adding a comment to this effect be helpful?

Reply:
If we are serious about controlling imports in `Init`, then maybe rather than adding a comment, we should actually think about what is wanted and how to enforce it.

Reply:
I agree this would be a good idea - and a follow-up PR. I just added a comment; feel free to tweak it.